File: ninchatsdk/src/main/java/com/ninchat/sdk/adapters/holders/formview/NinchatButtonViewHolder.java
package com.ninchat.sdk.adapters.holders.formview;

import android.os.Build;
import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;
import android.text.TextUtils;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.ninchat.sdk.NinchatSessionManager;
import com.ninchat.sdk.R;
import com.ninchat.sdk.events.OnNextQuestionnaire;

import org.greenrobot.eventbus.EventBus;
import org.json.JSONObject;

import static com.ninchat.sdk.helper.questionnaire.NinchatQuestionnaireMiscUtil.hasButton;
import static com.ninchat.sdk.helper.questionnaire.NinchatQuestionnaireMiscUtil.setAnimation;


public class NinchatButtonViewHolder extends RecyclerView.ViewHolder {
    private String TAG = NinchatButtonViewHolder.class.getSimpleName();

    private TextView mPrevious;
    private TextView mNext;
    private ImageView mPreviousImage;
    private ImageView mNextImage;

    public NinchatButtonViewHolder(@NonNull View itemView,
                                   JSONObject questionnaireElement,
                                   boolean isFormLikeQuestionnaire,
                                   int position) {
        super(itemView);
        mPrevious = itemView.findViewById(R.id.ninchat_button_previous);
        mNext = itemView.findViewById(R.id.ninchat_button_next);
        mPreviousImage = itemView.findViewById(R.id.ninchat_image_button_previous);
        mNextImage = itemView.findViewById(R.id.ninchat_image_button_next);
        this.bind(questionnaireElement, isFormLikeQuestionnaire, position, false);
    }

    public void bind(JSONObject questionnaireElement, boolean isFormLikeQuestionnaire, int position, boolean isUpdate) {
        mPrevious.setVisibility(View.GONE);
        mNext.setVisibility(View.GONE);
        mPreviousImage.setVisibility(View.GONE);
        mNextImage.setVisibility(View.GONE);
        if (hasButton(questionnaireElement, true)) {
            String text = questionnaireElement.optString("back");
            if ("true".equalsIgnoreCase(text) || TextUtils.isEmpty(text)) {
                mPreviousImage.setVisibility(View.VISIBLE);
                mPreviousImage.setOnClickListener(v -> {
                    mPreviousImage.setBackground(ContextCompat.getDrawable(itemView.getContext(), R.drawable.ninchat_chat_secondary_onclicked_button));
                    mayBeFireComplete(questionnaireElement, OnNextQuestionnaire.back);
                });
            } else {
                mPrevious.setVisibility(View.VISIBLE);
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    mPrevious.setTooltipText(NinchatSessionManager.getInstance().getTranslation(text));
                }

                mPrevious.setText(NinchatSessionManager.getInstance().getTranslation(text));
                mPrevious.setOnClickListener(v -> {
                    mPrevious.setBackground(ContextCompat.getDrawable(itemView.getContext(), R.drawable.ninchat_chat_secondary_onclicked_button));
                    mayBeFireComplete(questionnaireElement, OnNextQuestionnaire.back);
                });
            }
        }

        if (hasButton(questionnaireElement, false)) {
            String text = questionnaireElement.optString("next");
            if ("true".equalsIgnoreCase(text) || TextUtils.isEmpty(text)) {
                mNextImage.setVisibility(View.VISIBLE);
                mNextImage.setOnClickListener(v -> {
                    mNextImage.setBackground(ContextCompat.getDrawable(itemView.getContext(), R.drawable.ninchat_chat_primary_oncliked_button));
                    mayBeFireComplete(questionnaireElement, OnNextQuestionnaire.forward);
                });
            } else {
                mNext.setVisibility(View.VISIBLE);
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    mNext.setTooltipText(NinchatSessionManager.getInstance().getTranslation(text));
                }
                mNext.setText(NinchatSessionManager.getInstance().getTranslation(text));
                mNext.setOnClickListener(v -> {
                    mNext.setBackground(ContextCompat.getDrawable(itemView.getContext(), R.drawable.ninchat_chat_primary_oncliked_button));
                    mayBeFireComplete(questionnaireElement, OnNextQuestionnaire.forward);
                });
            }
        }

        if (!isUpdate)
            setAnimation(itemView, position, position != 0);
    }

    private void mayBeFireComplete(JSONObject questionnaireElement, int moveType) {
        if (questionnaireElement != null && questionnaireElement.optBoolean("fireEvent", false)) {
            if ("thankYouText".equalsIgnoreCase(questionnaireElement.optString("type", ""))) {
                EventBus.getDefault().post(new OnNextQuestionnaire(OnNextQuestionnaire.thankYou));
            } else {
                EventBus.getDefault().post(new OnNextQuestionnaire(moveType));
            }

        }
    }
}

File: ninchatsdk/src/main/java/com/ninchat/sdk/ninchatquestionnaire/view/NinchatButtonViewHolder.kt
package com.ninchat.sdk.ninchatquestionnaire.view

import android.view.View
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.ninchat.sdk.R
import com.ninchat.sdk.ninchatquestionnaire.model.NinchatButtonViewModel
import com.ninchat.sdk.ninchatquestionnaire.presenter.INinchatButtonViewPresenter
import com.ninchat.sdk.ninchatquestionnaire.presenter.NinchatButtonViewPresenter
import kotlinx.android.synthetic.main.control_buttons.view.*

class NinchatButtonViewHolder(
        itemView: View,
        ninchatButtonViewModel: NinchatButtonViewModel,
) : RecyclerView.ViewHolder(itemView), INinchatButtonViewPresenter {
    val ninchatButtonViewPresenter = NinchatButtonViewPresenter(
            ninchatButtonViewModel = ninchatButtonViewModel,
            iPresenter = this
    )

    init {
        ninchatButtonViewPresenter.renderCurrentView()
        attachUserActionHancler()
    }

    fun update(ninchatButtonViewModel: NinchatButtonViewModel) {
        ninchatButtonViewPresenter.run {
            updateViewModel(_ninchatButtonViewModel = ninchatButtonViewModel)
            renderCurrentView()
        }
        attachUserActionHancler()
    }

    private fun attachUserActionHancler() {
        // update background of the button
        // update background of the button
        itemView.run {
            ninchat_image_button_previous?.setOnClickListener {
                ninchatButtonViewPresenter.onBackButtonClicked()
            }
            ninchat_button_previous?.setOnClickListener {
                ninchatButtonViewPresenter.onBackButtonClicked()
            }
            ninchat_image_button_next?.setOnClickListener {
                ninchatButtonViewPresenter.onBackButtonClicked()
            }
            ninchat_button_next?.setOnClickListener {
                ninchatButtonViewPresenter.onBackButtonClicked()
            }
        }
    }

    override fun onBackButtonUpdated(visible: Boolean, text: String?, imageButton: Boolean, enabled: Boolean) {
        val background = if (enabled) R.drawable.ninchat_chat_primary_oncliked_button else R.drawable.ninchat_chat_secondary_button
        itemView.run {
            if (imageButton) {
                ninchat_image_button_previous.background = ContextCompat.getDrawable(itemView.context, background)
                ninchat_image_button_previous?.visibility = if (visible) View.VISIBLE else View.GONE
            } else {
                ninchat_button_previous?.background = ContextCompat.getDrawable(itemView.context, background)
                ninchat_button_previous?.visibility = if (visible) View.VISIBLE else View.GONE
                ninchat_button_previous?.text = text
            }
        }
    }

    override fun onNextNextUpdated(visible: Boolean, text: String?, imageButton: Boolean, enabled: Boolean) {
        val background = if (enabled) R.drawable.ninchat_chat_primary_oncliked_button else R.drawable.ninchat_chat_secondary_button
        itemView.run {
            if (imageButton) {
                ninchat_image_button_next?.background = ContextCompat.getDrawable(itemView.context, background)
                ninchat_image_button_next?.visibility = if (visible) View.VISIBLE else View.GONE
            } else {
                ninchat_button_next?.background = ContextCompat.getDrawable(itemView.context, background)
                ninchat_button_next?.visibility = if (visible) View.VISIBLE else View.GONE
                ninchat_button_next?.text = text
            }
        }
    }
}

interface INinchatButtonViewHolder {
    fun onBackButtonClicked()
    fun onNextButtonClicked()
}